File: packages/arcis-python/arcis/guards.py
"""
Guards API for Python.

Same Arcis decisioning (rate limit, bot detect, prompt injection, token
budget) applied to non-HTTP contexts where there is no request/response
pair. Use this for job queue workers, agent tool-call handlers,
WebSocket / SSE / gRPC handlers, background processors.

Mirrors the Node implementation at ``packages/arcis-node/src/guards.ts``.

Example:

.. code-block:: python

    from arcis.guards import Guards

    guards = Guards(
        rate_limit={"max": 50, "window_ms": 60_000},
        token_budget={"max_tokens": 100_000, "window_ms": 3_600_000},
        prompt_injection={"deny_at": "medium"},
        bot={"deny": ["AUTOMATED"]},
    )

    # In a Celery task or queue worker:
    decision = guards.run(
        key=job_user_id,
        text=user_prompt,
        tokens=estimate_tokens(user_prompt),
    )
    if not decision.ok:
        raise GuardDenied(f"{decision.vector}: {decision.reason}")
"""

from __future__ import annotations

import threading
import time
from dataclasses import dataclass, field
from typing import Any, List, Literal, Optional, Tuple, Union

from .middleware.bot_detection import detect_bot
from .sanitizers.prompt_injection import detect_prompt_injection


GuardsVector = Literal["rate-limit", "token-budget", "prompt-injection", "bot"]
GuardsSeverity = Literal["low", "medium", "high"]


@dataclass
class GuardsDecision:
    """Result of evaluating a guards input.

    `ok` is True when every configured vector passes. When False, `vector`
    names the first denying vector and the rest of the fields describe it.
    `matches` carries prompt-injection signature matches even when the
    deny threshold was not crossed, so callers can log low-severity hits
    without blocking on them.
    """
    ok: bool
    vector: Optional[GuardsVector] = None
    reason: Optional[str] = None
    severity: Optional[GuardsSeverity] = None
    retry_after_seconds: Optional[int] = None
    matches: List[Tuple[str, GuardsSeverity]] = field(default_factory=list)


@dataclass
class _RLEntry:
    count: int
    reset_time: float


@dataclass
class _TBEntry:
    used: int
    reset_time: float


_SEVERITY_RANK = {"low": 1, "medium": 2, "high": 3}
_DEFAULT_BOT_ALLOW = frozenset({"SEARCH_ENGINE", "SOCIAL", "MONITORING"})
_DEFAULT_BOT_DENY = frozenset({"AUTOMATED"})


def _coerce(option: Any) -> dict:
    """Treat ``True`` as ``{}`` (use defaults) and dicts as themselves.
    Anything else means the vector is disabled."""
    if option is True:
        return {}
    if isinstance(option, dict):
        return option
    return {}


class Guards:
    """Per-instance state for rate-limit and token-budget buckets.

    Construct once with the vectors you care about, then call
    :meth:`run` per request/event. Thread-safe in-memory store. Call
    :meth:`close` to release the periodic-cleanup background thread.
    """

    def __init__(
        self,
        rate_limit: Optional[Union[dict, bool]] = None,
        token_budget: Optional[Union[dict, bool]] = None,
        prompt_injection: Optional[Union[dict, bool]] = None,
        bot: Optional[Union[dict, bool]] = None,
    ):
        self._rl = _coerce(rate_limit) if rate_limit else None
        self._tb = _coerce(token_budget) if token_budget else None
        self._pi = _coerce(prompt_injection) if prompt_injection else None
        self._bot = _coerce(bot) if bot else None

        deny_at = (self._pi or {}).get("deny_at", "medium")
        self._pi_deny_rank = _SEVERITY_RANK[deny_at]

        self._rl_store: dict[str, _RLEntry] = {}
        self._tb_store: dict[str, _TBEntry] = {}
        self._lock = threading.Lock()

        # Background cleanup only when a time-windowed vector is configured.
        self._stop_cleanup: Optional[threading.Event] = None
        self._cleanup_thread: Optional[threading.Thread] = None
        if self._rl or self._tb:
            window_ms = (
                (self._rl or {}).get("window_ms")
                or (self._tb or {}).get("window_ms")
                or 60_000
            )
            self._stop_cleanup = threading.Event()
            self._cleanup_thread = threading.Thread(
                target=self._sweep_loop,
                args=(window_ms / 1000.0,),
                daemon=True,
            )
            self._cleanup_thread.start()

    def run(
        self,
        key: str,
        text: Optional[str] = None,
        tokens: Optional[int] = None,
        user_agent: Optional[str] = None,
    ) -> GuardsDecision:
        """Evaluate every configured vector against the input.

        Returns a :class:`GuardsDecision`. The first denying vector
        short-circuits the rest, so a prompt-injection deny does not
        charge the per-key token budget.
        """
        if not isinstance(key, str) or not key:
            return GuardsDecision(ok=False, reason="guards: missing required `key`")

        # 1. Rate limit
        if self._rl is not None:
            decision = self._check_rate_limit(key)
            if not decision.ok:
                return decision

        # 2. Bot detection (only when a UA was supplied)
        if self._bot is not None and user_agent:
            decision = self._check_bot(user_agent)
            if not decision.ok:
                return decision

        # 3. Prompt injection (only when text was supplied)
        pi_matches: List[Tuple[str, GuardsSeverity]] = []
        if self._pi is not None and isinstance(text, str) and text:
            r = detect_prompt_injection(text)
            pi_matches = [(m.rule, m.severity) for m in r.matches]
            if (
                r.detected
                and r.severity != "none"
                and _SEVERITY_RANK[r.severity] >= self._pi_deny_rank
            ):
                top = next((m for m in r.matches if m.severity == r.severity), None)
                return GuardsDecision(
                    ok=False,
                    vector="prompt-injection",
                    severity=r.severity,
                    reason=(
                        f"Prompt injection detected ({top.rule}): {top.description}"
                        if top else "Prompt injection detected"
                    ),
                    matches=pi_matches,
                )

        # 4. Token budget
        if self._tb is not None and isinstance(tokens, (int, float)):
            decision = self._check_token_budget(key, tokens)
            if not decision.ok:
                decision.matches = pi_matches
                return decision

        return GuardsDecision(ok=True, matches=pi_matches)

    def inspect_rate_limit(self, key: str) -> Optional[dict]:
        with self._lock:
            e = self._rl_store.get(key)
            return {"count": e.count, "reset_time": e.reset_time} if e else None

    def inspect_token_budget(self, key: str) -> Optional[dict]:
        with self._lock:
            e = self._tb_store.get(key)
            return {"used": e.used, "reset_time": e.reset_time} if e else None

    def reset(self, key: Optional[str] = None) -> None:
        """Clear a single key's state, or every key if ``key`` is None."""
        with self._lock:
            if key is None:
                self._rl_store.clear()
                self._tb_store.clear()
            else:
                self._rl_store.pop(key, None)
                self._tb_store.pop(key, None)

    def close(self) -> None:
        """Stop the background cleanup thread. Idempotent."""
        if self._stop_cleanup is not None and self._cleanup_thread is not None:
            self._stop_cleanup.set()
            self._cleanup_thread.join(timeout=0.1)

    # ── internals ────────────────────────────────────────────────────────

    def _check_rate_limit(self, key: str) -> GuardsDecision:
        cfg = self._rl or {}
        max_n = int(cfg.get("max", 100))
        window_ms = int(cfg.get("window_ms", 60_000))
        now = time.monotonic()
        with self._lock:
            entry = self._rl_store.get(key)
            if entry is None or entry.reset_time < now:
                entry = _RLEntry(count=0, reset_time=now + window_ms / 1000.0)
                self._rl_store[key] = entry
            entry.count += 1
            if entry.count > max_n:
                retry = max(0, int(entry.reset_time - now + 0.999))
                return GuardsDecision(
                    ok=False,
                    vector="rate-limit",
                    severity="medium",
                    reason=f"Rate limit exceeded ({entry.count}/{max_n} per {window_ms}ms)",
                    retry_after_seconds=retry,
                )
        return GuardsDecision(ok=True)

    def _check_token_budget(self, key: str, tokens: float) -> GuardsDecision:
        cfg = self._tb or {}
        cost = max(0, int(tokens)) if isinstance(tokens, (int, float)) and tokens == tokens else 0
        max_t = int(cfg.get("max_tokens", 100_000))
        window_ms = int(cfg.get("window_ms", 60 * 60 * 1000))
        per_req = cfg.get("max_request_tokens")

        if per_req is not None and cost > int(per_req):
            return GuardsDecision(
                ok=False,
                vector="token-budget",
                severity="high",
                reason=f"Per-call token budget exceeded ({cost} > {int(per_req)})",
            )

        now = time.monotonic()
        with self._lock:
            entry = self._tb_store.get(key)
            if entry is None or entry.reset_time < now:
                entry = _TBEntry(used=0, reset_time=now + window_ms / 1000.0)
                self._tb_store[key] = entry
            projected = entry.used + cost
            if projected > max_t:
                retry = max(0, int(entry.reset_time - now + 0.999))
                return GuardsDecision(
                    ok=False,
                    vector="token-budget",
                    severity="medium",
                    reason=f"Window token budget exceeded ({entry.used} + {cost} > {max_t})",
                    retry_after_seconds=retry,
                )
            entry.used = projected
        return GuardsDecision(ok=True)

    def _check_bot(self, user_agent: str) -> GuardsDecision:
        # Build a request-shaped object detect_bot can read off without
        # importing a web framework.
        class _Req:
            def __init__(self, ua: str):
                self.headers = {
                    "user-agent": ua,
                    "accept": "text/html",
                    "accept-language": "en-US",
                    "accept-encoding": "gzip",
                }
        result = detect_bot(_Req(user_agent))
        if not result.is_bot:
            return GuardsDecision(ok=True)

        cfg = self._bot or {}
        allow = set(cfg.get("allow")) if cfg.get("allow") else _DEFAULT_BOT_ALLOW
        deny = set(cfg.get("deny")) if cfg.get("deny") else _DEFAULT_BOT_DENY
        default_action = cfg.get("default_action", "allow")

        if result.category in allow:
            return GuardsDecision(ok=True)
        if result.category in deny:
            return GuardsDecision(
                ok=False,
                vector="bot",
                severity="medium",
                reason=(
                    f"Bot denied: {result.name}" if result.name
                    else f"Bot denied ({result.category})"
                ),
            )
        if default_action == "deny":
            return GuardsDecision(
                ok=False,
                vector="bot",
                severity="low",
                reason="Uncategorized bot under default_action=deny",
            )
        return GuardsDecision(ok=True)

    def _sweep_loop(self, interval_seconds: float) -> None:
        assert self._stop_cleanup is not None
        while not self._stop_cleanup.wait(interval_seconds):
            now = time.monotonic()
            with self._lock:
                for store in (self._rl_store, self._tb_store):
                    expired = [k for k, e in store.items() if e.reset_time < now]
                    for k in expired:
                        del store[k]


__all__ = [
    "Guards",
    "GuardsDecision",
    "GuardsVector",
    "GuardsSeverity",
]

File: packages/arcis-node/src/guards.ts
/**
 * @module @arcis/node/guards
 *
 * Guards API. Same Arcis decisioning (rate limit, bot detect, prompt
 * injection, token budget) applied to non-HTTP contexts where there's no
 * `req`/`res` pair. Use this for:
 *
 *   - Job queue workers (BullMQ, agenda, sidekiq-style)
 *   - Agent tool-call handlers (Claude/OpenAI tool dispatch)
 *   - WebSocket / SSE / gRPC handlers
 *   - Background processors (cron jobs, scheduled tasks)
 *
 * Each call to `guards.run(input)` returns a structured decision: `ok` +
 * (when denied) the `vector`, `severity`, `reason`, and `retryAfterSeconds`
 * the deny was triggered by. The first vector that denies short-circuits
 * the rest, so denial latency stays bounded.
 *
 * @example
 * import { Guards } from '@arcis/node';
 *
 * const guards = new Guards({
 *   rateLimit: { max: 50, windowMs: 60_000 },
 *   tokenBudget: { maxTokens: 100_000, windowMs: 60 * 60 * 1000 },
 *   promptInjection: { redactLow: false },
 * });
 *
 * // In a job handler:
 * const decision = guards.run({
 *   key: jobUserId,
 *   tokens: estimateTokens(prompt),
 *   text: prompt,
 * });
 * if (!decision.ok) {
 *   throw new Error(`Job rejected (${decision.vector}): ${decision.reason}`);
 * }
 */

import { detectBot, type BotProtectionOptions } from './middleware/bot-detection';
import {
  detectPromptInjection,
  type PromptInjectionSeverity,
} from './sanitizers/prompt-injection';

// ─── Public types ──────────────────────────────────────────────────────────

export interface GuardsRateLimitOptions {
  /** Max events per window per key. Default: 100. */
  max?: number;
  /** Window length in milliseconds. Default: 60000 (1 minute). */
  windowMs?: number;
}

export interface GuardsTokenBudgetOptions {
  /** Max tokens a single key can spend in one window. Default: 100,000. */
  maxTokens?: number;
  /** Window length in milliseconds. Default: 60 * 60 * 1000 (1 hour). */
  windowMs?: number;
  /**
   * Optional per-call cap. When set, a single call with `tokens > maxRequestTokens`
   * denies BEFORE charging the window budget.
   */
  maxRequestTokens?: number;
}

export interface GuardsPromptInjectionOptions {
  /**
   * Minimum severity that triggers a deny. Default: 'medium' (HIGH and
   * MEDIUM matches deny; LOW matches still surface in `decision.matches`
   * but don't deny).
   */
  denyAt?: PromptInjectionSeverity;
}

export interface GuardsBotOptions {
  /** Categories that pass through. Default: SEARCH_ENGINE, SOCIAL, MONITORING. */
  allow?: BotProtectionOptions['allow'];
  /** Categories that always deny. Default: AUTOMATED. */
  deny?: BotProtectionOptions['deny'];
  /** Default for uncategorized bots. Default: 'allow'. */
  defaultAction?: BotProtectionOptions['defaultAction'];
}

export interface GuardsConfig {
  /** When set, every call is rate-limited per `input.key`. Omit to disable. */
  rateLimit?: GuardsRateLimitOptions;
  /** When set, calls with `input.tokens` charge a per-key sliding-window budget. */
  tokenBudget?: GuardsTokenBudgetOptions;
  /** When set, `input.text` is scanned for prompt-injection signatures. */
  promptInjection?: GuardsPromptInjectionOptions | true;
  /** When set, `input.userAgent` is matched against the bot corpus. */
  bot?: GuardsBotOptions | true;
}

export interface GuardsInput {
  /** Identifier for rate-limit / token-budget bucketing. Required. */
  key: string;
  /** Optional text payload for prompt-injection scanning. */
  text?: string;
  /** Optional token cost for token-budget accounting. */
  tokens?: number;
  /** Optional User-Agent string for bot detection. */
  userAgent?: string;
}

export type GuardsVector =
  | 'rate-limit'
  | 'token-budget'
  | 'prompt-injection'
  | 'bot';

export type GuardsSeverity = 'low' | 'medium' | 'high';

export interface GuardsDecision {
  /** True if the input passes every configured vector. */
  ok: boolean;
  /** Which vector denied. Undefined when `ok` is true. */
  vector?: GuardsVector;
  /** Human-readable reason for the deny. Undefined when `ok` is true. */
  reason?: string;
  /** Severity of the deny. Undefined when `ok` is true. */
  severity?: GuardsSeverity;
  /** How many seconds until the same key can retry (rate-limit / token-budget). */
  retryAfterSeconds?: number;
  /**
   * For prompt-injection: every signature that matched, even when the deny
   * threshold wasn't hit. Lets callers log low-severity matches without
   * blocking on them.
   */
  matches?: ReadonlyArray<{ rule: string; severity: GuardsSeverity }>;
}

// ─── Internal state ────────────────────────────────────────────────────────

interface RLEntry {
  count: number;
  resetTime: number;
}

interface TBEntry {
  used: number;
  resetTime: number;
}

const SEVERITY_RANK: Record<GuardsSeverity, number> = { low: 1, medium: 2, high: 3 };

const DEFAULT_BOT_ALLOW = new Set(['SEARCH_ENGINE', 'SOCIAL', 'MONITORING']);
const DEFAULT_BOT_DENY = new Set(['AUTOMATED']);

// ─── Public class ──────────────────────────────────────────────────────────

/**
 * Guards. Apply Arcis decisions to non-HTTP contexts. Construct once with
 * the vectors you care about, then call `.run(input)` per request/event.
 * Internal state (rate-limit buckets, token-budget buckets) lives on the
 * instance. Call `.close()` to release the periodic-cleanup interval.
 */
export class Guards {
  private readonly rl: GuardsRateLimitOptions | undefined;
  private readonly tb: GuardsTokenBudgetOptions | undefined;
  private readonly pi: GuardsPromptInjectionOptions | undefined;
  private readonly bot: GuardsBotOptions | undefined;
  private readonly rlStore: Record<string, RLEntry>;
  private readonly tbStore: Record<string, TBEntry>;
  private readonly cleanup: ReturnType<typeof setInterval> | null;
  private readonly piDenyRank: number;

  constructor(config: GuardsConfig = {}) {
    this.rl = config.rateLimit;
    this.tb = config.tokenBudget;
    this.pi = config.promptInjection === true ? {} : config.promptInjection;
    this.bot = config.bot === true ? {} : config.bot;

    const denyAt = (this.pi?.denyAt as GuardsSeverity | undefined) ?? 'medium';
    this.piDenyRank = SEVERITY_RANK[denyAt];

    this.rlStore = Object.create(null);
    this.tbStore = Object.create(null);

    // Sweep expired buckets only when rate-limit or token-budget is in play.
    const sweepInterval = this.rl?.windowMs ?? this.tb?.windowMs;
    if (sweepInterval) {
      this.cleanup = setInterval(() => this.sweepExpired(), sweepInterval);
      if (typeof this.cleanup.unref === 'function') this.cleanup.unref();
    } else {
      this.cleanup = null;
    }
  }

  /**
   * Evaluate every configured vector against `input`. Returns a structured
   * decision; the first denying vector short-circuits the rest.
   */
  run(input: GuardsInput): GuardsDecision {
    if (!input || typeof input.key !== 'string' || input.key.length === 0) {
      return { ok: false, reason: 'guards: missing required `input.key`' };
    }

    // 1. Rate limit
    if (this.rl) {
      const decision = this.checkRateLimit(input.key);
      if (!decision.ok) return decision;
    }

    // 2. Bot detection (if a UA was supplied)
    if (this.bot && input.userAgent) {
      const decision = this.checkBot(input.userAgent);
      if (!decision.ok) return decision;
    }

    // 3. Prompt injection (if text was supplied)
    let piMatches: ReadonlyArray<{ rule: string; severity: GuardsSeverity }> | undefined;
    if (this.pi !== undefined && typeof input.text === 'string' && input.text.length > 0) {
      const result = detectPromptInjection(input.text);
      piMatches = result.matches.map((m) => ({ rule: m.rule, severity: m.severity }));
      if (
        result.detected &&
        result.severity !== 'none' &&
        SEVERITY_RANK[result.severity] >= this.piDenyRank
      ) {
        const top = result.matches.find((m) => m.severity === result.severity)!;
        return {
          ok: false,
          vector: 'prompt-injection',
          severity: result.severity,
          reason: `Prompt injection detected (${top.rule}): ${top.description}`,
          matches: piMatches,
        };
      }
    }

    // 4. Token budget (always last so a denied request hasn't already charged)
    if (this.tb && typeof input.tokens === 'number') {
      const decision = this.checkTokenBudget(input.key, input.tokens);
      if (!decision.ok) return { ...decision, matches: piMatches };
    }

    return { ok: true, matches: piMatches };
  }

  /** Inspect rate-limit usage for a key. Useful for tests and telemetry. */
  inspectRateLimit(key: string): { count: number; resetTime: number } | null {
    const e = this.rlStore[key];
    return e ? { count: e.count, resetTime: e.resetTime } : null;
  }

  /** Inspect token-budget usage for a key. */
  inspectTokenBudget(key: string): { used: number; resetTime: number } | null {
    const e = this.tbStore[key];
    return e ? { used: e.used, resetTime: e.resetTime } : null;
  }

  /** Reset a single key's state, or all keys if `key` is omitted. */
  reset(key?: string): void {
    if (key === undefined) {
      for (const k of Object.keys(this.rlStore)) delete this.rlStore[k];
      for (const k of Object.keys(this.tbStore)) delete this.tbStore[k];
    } else {
      delete this.rlStore[key];
      delete this.tbStore[key];
    }
  }

  /** Release the periodic cleanup interval. Idempotent. */
  close(): void {
    if (this.cleanup !== null) clearInterval(this.cleanup);
  }

  // ─── internals ────────────────────────────────────────────────────────

  private checkRateLimit(key: string): GuardsDecision {
    const max = this.rl?.max ?? 100;
    const windowMs = this.rl?.windowMs ?? 60_000;
    const now = Date.now();
    let entry = this.rlStore[key];
    if (!entry || entry.resetTime < now) {
      entry = { count: 0, resetTime: now + windowMs };
      this.rlStore[key] = entry;
    }
    entry.count += 1;
    if (entry.count > max) {
      const retryAfterSeconds = Math.ceil((entry.resetTime - now) / 1000);
      return {
        ok: false,
        vector: 'rate-limit',
        severity: 'medium',
        reason: `Rate limit exceeded (${entry.count}/${max} per ${windowMs}ms)`,
        retryAfterSeconds: Math.max(0, retryAfterSeconds),
      };
    }
    return { ok: true };
  }

  private checkTokenBudget(key: string, tokens: number): GuardsDecision {
    const cost = Number.isFinite(tokens) && tokens >= 0 ? Math.floor(tokens) : 0;
    const max = this.tb?.maxTokens ?? 100_000;
    const windowMs = this.tb?.windowMs ?? 60 * 60 * 1000;
    const perReq = this.tb?.maxRequestTokens;

    if (perReq !== undefined && cost > perReq) {
      return {
        ok: false,
        vector: 'token-budget',
        severity: 'high',
        reason: `Per-call token budget exceeded (${cost} > ${perReq})`,
      };
    }

    const now = Date.now();
    let entry = this.tbStore[key];
    if (!entry || entry.resetTime < now) {
      entry = { used: 0, resetTime: now + windowMs };
      this.tbStore[key] = entry;
    }
    const projected = entry.used + cost;
    if (projected > max) {
      const retryAfterSeconds = Math.ceil((entry.resetTime - now) / 1000);
      return {
        ok: false,
        vector: 'token-budget',
        severity: 'medium',
        reason: `Window token budget exceeded (${entry.used} + ${cost} > ${max})`,
        retryAfterSeconds: Math.max(0, retryAfterSeconds),
      };
    }
    entry.used = projected;
    return { ok: true };
  }

  private checkBot(userAgent: string): GuardsDecision {
    const fakeReq = {
      headers: {
        'user-agent': userAgent,
        accept: 'text/html',
        'accept-language': 'en-US',
        'accept-encoding': 'gzip',
      },
    };
    const result = detectBot(fakeReq as never);
    if (!result.isBot) return { ok: true };

    const allow = this.bot?.allow ? new Set(this.bot.allow) : DEFAULT_BOT_ALLOW;
    const deny = this.bot?.deny ? new Set(this.bot.deny) : DEFAULT_BOT_DENY;
    const defaultAction = this.bot?.defaultAction ?? 'allow';

    if (allow.has(result.category as never)) return { ok: true };
    if (deny.has(result.category as never)) {
      return {
        ok: false,
        vector: 'bot',
        severity: 'medium',
        reason: result.name ? `Bot denied: ${result.name}` : `Bot denied (${result.category})`,
      };
    }
    if (defaultAction === 'deny') {
      return {
        ok: false,
        vector: 'bot',
        severity: 'low',
        reason: `Uncategorized bot under defaultAction=deny`,
      };
    }
    return { ok: true };
  }

  private sweepExpired(): void {
    const now = Date.now();
    for (const k of Object.keys(this.rlStore)) {
      if (this.rlStore[k].resetTime < now) delete this.rlStore[k];
    }
    for (const k of Object.keys(this.tbStore)) {
      if (this.tbStore[k].resetTime < now) delete this.tbStore[k];
    }
  }
}

export default Guards;
